<a href="https://colab.research.google.com/github/vinalyan/diff/blob/master/%D0%9C%D0%B0%D1%82%D1%81%D1%82%D0%B0%D1%82_%D1%83%D1%80%D0%BE%D0%BA_6_%D0%94%D0%97_%D0%9A%D0%BE%D0%B6%D0%B5%D0%BC%D1%8F%D0%BA%D0%B8%D0%BD_%D0%92_%D0%95.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Настройка


In [55]:
import numpy as np
from math import factorial
def combinations(n, k):
    return int(factorial(n) / (factorial(k) * factorial(n - k)))

from matplotlib import pyplot as plt
import seaborn as sns
from scipy.stats import norm
import pandas as pd

print('Настройка завершена')

Настройка завершена


#Задание №1
Даны значения величины заработной платы заемщиков банка (zp) и значения их поведенческого кредитного скоринга (ks):
zp = [35, 45, 190, 200, 40, 70, 54, 150, 120, 110],
ks = [401, 574, 874, 919, 459, 739, 653, 902, 746, 832].
Найдите ковариацию этих двух величин с помощью элементарных действий, а затем с помощью функции cov из numpy.
Полученные значения должны быть равны.
Найдите коэффициент корреляции Пирсона с помощью ковариации и среднеквадратичных отклонений двух признаков,
а затем с использованием функций из библиотек numpy и pandas.

## Решение
Найдем ковариацию этих двух волшебных величичию. Ковариация определяется по формуле:
$$cov_{XY} = M(XY) - M(X)M(Y)$$

Коэффициент корреляции определяется по формуе:

$$r_{XY} = \frac {cov_{XY}}{\sigma_X \sigma_Y}$$

In [56]:
zp = np.array([35, 45, 190, 200, 40, 70, 54, 150, 120, 110])
ks = np.array([401, 574, 874, 919, 459, 739, 653, 902, 746, 832])
s = zp * ks

In [57]:
#Считаем ковариацию практически руками
cov = s.mean() - zp.mean() * ks.mean()
cov

9157.839999999997

In [58]:
#Считаем ковариацию c помощью функции cov 
np.cov( zp , ks, ddof = 0 )

array([[ 3494.64,  9157.84],
       [ 9157.84, 30468.89]])

In [59]:
#Практически руками считаем коэфициент корреляции
sigma_zp = np.std(zp)
sigma_ks = np.std(ks)
r= cov / (sigma_zp * sigma_ks)
r

0.8874900920739158

In [60]:
#Считаем корреляцию с помощью функции из numpy
np.corrcoef(zp , ks)

array([[1.        , 0.88749009],
       [0.88749009, 1.        ]])

In [61]:
#Тренируемся с пандами
df = pd.DataFrame({'ks': ks ,'zp': zp})
df

ks   zp
0  401   35
1  574   45
2  874  190
3  919  200
4  459   40
5  739   70
6  653   54
7  902  150
8  746  120
9  832  110

In [62]:
df.corr()

ks       zp
ks  1.00000  0.88749
zp  0.88749  1.00000

##Ответ
$cov_{XY} = 9157$

$r_{XY} = 8,89$ 

#Задание №2
Измерены значения IQ выборки студентов,
обучающихся в местных технических вузах:
131, 125, 115, 122, 131, 115, 107, 99, 125, 111.
Известно, что в генеральной совокупности IQ распределен нормально.
Найдите доверительный интервал для математического ожидания с надежностью 0.95.

##Решение

Так как СКО генеральной совокупности неизвестно, то будет использовать $t$ критерий Стьюдента.
$$\overline X \pm t_{a/2}* \sigma / \sqrt n$$
где: 

$t_{a/2} = 2,3$ - определяем по таблице Стьюдента с надежностью $0,95$ и степенями свободы равным $9$

$\overline X = 118,1$

$\sigma = 10,5$ - несмещенное СКО выборки


In [63]:
#Вычисляем математическое ожидаени
iq = np.array([131, 125, 115, 122, 131, 115, 107, 99, 125, 111])
X = iq.mean()
t = 2.3
#Вычисляем несмещенное СКО
sigma = np.std(iq , ddof = 1)

#Вычисляем доверительный интервал
l = X - t * (sigma/(iq.size**0.5))
r = X + t* (sigma/(iq.size**0.5))

print('Среднее арифметическое: {}'.format(X))
print('Несмещенное СКО: {:.3}'.format(sigma))
print('Доверительный интервал: [{:.6};{:.6}]'.format(l,r))

Среднее арифметическое: 118.1
Несмещенное СКО: 10.5
Доверительный интервал: [110.43;125.77]


##Ответ
Доверительный интервал: $[110.43;125.77]$

#Задание №3
Известно, что рост футболистов в сборной распределен нормально
с дисперсией генеральной совокупности, равной 25 кв.см. Объем выборки равен 27,
среднее выборочное составляет 174.2. Найдите доверительный интервал для математического
ожидания с надежностью 0.95.

##Решение
Так как $\sigma$ - СКО генеральной совокупности известно, то будем использовать $Z$ критерий.

Доверительный интервал определяется по следующей формуле:
$$\overline X \pm Z_{a/2}* \sigma / \sqrt n$$
где: 

$Z_{a/2} = 1,96$

$\overline X = 174,2$

$\sigma = 5$

$n = 27$

Подставив все это в формулу получим доверительный интервал для математического ожидания $[172.3 ; 176.1]$
 

In [64]:
z = norm.ppf(0.975)
sigma = 5
n = 27
X = 174.2
l = X - z * (sigma/(n**0.5))
r = X + z* (sigma/(n**0.5))


Доверительный интервал: [172.3 ; 176.1]


##Ответ
Доверительный интервал:

$[172.3 ; 176.1]$

In [66]:
print('Доверительный интервал: [{:.4} ; {:.4}]'.format(l,r))

Доверительный интервал: [172.3 ; 176.1]


In [65]:
z = norm.ppf(0.025)
z

-1.9599639845400545